environments/common/src/main/java/org/jboss/weld/environment/deployment/discovery/jandex/JandexClassFileServices.java

```diff
@@ -74,10 +77,20 @@ public JandexClassFileServices(JandexDiscoveryStrategy strategy) {
         }
         this.classLoader = WeldResourceLoader.getClassLoader();
         this.annotationClassAnnotationsCache = ComputingCacheBuilder.newBuilder().build(new AnnotationClassAnnotationLoader());
+        try {
+            Method setFlags = ClassFileInfo.class.getMethod("setFlags");
+            jandexStaticCheckAvailable = setFlags != null;
+        } catch (NoSuchMethodException e) {
+            jandexStaticCheckAvailable = false;
+            CommonLogger.LOG.usingOldJandexVersion();
+        }
     }
 
     @Override
     public ClassFileInfo getClassFileInfo(String className) {
+        if (!jandexStaticCheckAvailable) {
+            throw new ClassFileInfoException("Using old Jandex version.");
+        }
```

Comment:
Altough it seems logical to throw exception here, I'd rather throw it in `ClassFileInfo.getModifiers()`. The reason is performance - `FastAnnotatedTypeLoader` can sucessfully "skip" many classes (annotation, `@Vetoed`... see also `FastAnnotatedTypeLoader.loadAnnotatedType()`)